src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -348,7 +329,7 @@ private static Project CreateProject(ProjectId projectId, Solution solution)
     }
 
     private Solution WithCompilationState(SolutionCompilationState compilationState)
-        => compilationState == _compilationState ? this : new Solution(compilationState);
+        => compilationState == _compilationState ? this : new Solution(compilationState, _projectIdMapping.WithProjectIds(compilationState.SolutionState.ProjectIds));
```

Comment:
since you're jsut accessing SolutionState.ProjectIds here, i think it would be best to just do the reference-equality check within the ProjectIdMap.  99.9999% of the time that will be true, and we'll reuse the frozen. dict.  And in the 0.001% when it's false, doing a sequence equals won't be beneficial.